src/node/txreconciliation.cpp

```diff
@@ -142,9 +250,104 @@ class TxReconciliationTracker::Impl
         return (recon_state != m_states.end() &&
                 std::holds_alternative<TxReconciliationState>(recon_state->second));
     }
+
+    // Not const because of caching.
+    bool IsFanoutTarget(const Wtxid& wtxid, NodeId peer_id, bool we_initiate, double limit) EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
+    {
+        auto fanout_candidates = m_tx_fanout_targets_cache_data.find(wtxid);
+        if (fanout_candidates != m_tx_fanout_targets_cache_data.end()) {
+            return std::binary_search(fanout_candidates->second.begin(), fanout_candidates->second.end(), peer_id);
+        }
+
+        // We use the pre-determined randomness to give a consistent result per transaction,
+        // thus making sure that no transaction gets "unlucky" if every per-peer roll fails.
+        CSipHasher deterministic_randomizer{m_deterministic_randomizer};
+        deterministic_randomizer.Write(wtxid.ToUint256());
+
+        // The algorithm works as follows. We iterate through the peers (of a given direction)
+        // hashing them with the given wtxid, and sort them by this hash.
+        // We then consider top `limit` peers to be low-fanout flood targets.
+        // The randomness should be seeded with wtxid to return consistent results for every call.
+
+        // The integral part of `limit` is accounted in the higher 32 bits of the second element.
+        // The fractional part of `limit` is stored in the lower 32 bits, and then we check
+        // whether adding a random lower 32-bit value (first element) would end up modifying
+        // the higher bits.
+        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(limit * 0x100000000)) >> 32;
+
+        std::vector<std::pair<uint64_t, NodeId>> best_peers;
+        best_peers.reserve(m_states.size());
+
+        for (const auto& indexed_state : m_states) {
+            const auto cur_state = std::get_if<TxReconciliationState>(&indexed_state.second);
+            if (cur_state && cur_state->m_we_initiate == we_initiate) {
+                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(cur_state->m_k0).Finalize();
```

Comment:
I think the comment L66 in `src/node/txreconciliation.cpp` can be updated to reflect the usage of `m_k0` as a siphash input string for low-fanout flood peers selection. Not only used in `ComputeShortID`.

Reply:
I think this can actually be seeded with anything, it doesn't have to be `m_k0`. IMO it'd better not be, to not repurpose something that is meant for something completely different